File: src/0783_minimum_distance_between_bst_nodes.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * DFS (level order traversal)
 * Time O(n) | Space O(n)
 * https://leetcode.com/problems/minimum-distance-between-bst-nodes/
 * @param {TreeNode} root
 * @return {number}
 */
var minDiffInBST = function (root) {
    // levelOrderTraversal
    const sortedArr = dfs(root, []);

    let min = Infinity;
    for (let i = 1; i < sortedArr.length; i++) {
        min = Math.min(min, sortedArr[i] - sortedArr[i - 1]);
    }
    return min;
};

const dfs = (node, sortedArr) => {
    if (!node) return;

    dfs(node.left, sortedArr);
    sortedArr.push(node.val);
    dfs(node.right, sortedArr);

    return sortedArr;
};

var minDiffInBST = function (root) {
    let currSmallest = Infinity;
    let prev = null;

    const inorder = (node) => {
        if (node === null) {
            return;
        }
        inorder(node.left);
        if (prev !== null) {
            currSmallest = Math.min(currSmallest, node.val - prev.val);
        }
        prev = node;
        inorder(node.right);
    };

    inorder(root);
    return currSmallest;
};

module.exports = { minDiffInBST };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(minDiffInBST(new TreeNode(4, new TreeNode(2, new TreeNode(1, null, null), new TreeNode(3, null, null)), new TreeNode(6, null, null))), 1);
    assert.deepStrictEqual(minDiffInBST(new TreeNode(1, new TreeNode(0, null, null), new TreeNode(48, new TreeNode(12, null, null), new TreeNode(49, null, null)))), 1);
}

File: src/0783_minimum_distance_between_bst_nodes.rs
use std::cell::RefCell;
use std::rc::Rc;

struct Solution;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        Self {
            val,
            left: None,
            right: None,
        }
    }
}

impl Solution {
    pub fn min_diff_in_bst(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        fn inorder(
            node: Option<Rc<RefCell<TreeNode>>>,
            curr_smallest: &mut i32,
            prev: &mut Option<i32>,
        ) {
            let Some(node) = node else {
                return;
            };

            let borrowed = node.borrow();
            inorder(borrowed.left.clone(), curr_smallest, prev);
            if let Some(prev_val) = *prev {
                *curr_smallest = (*curr_smallest).min(borrowed.val - prev_val);
            }
            *prev = Some(borrowed.val);
            inorder(borrowed.right.clone(), curr_smallest, prev);
        }

        let mut curr_smallest = i32::MAX;
        let mut prev = None;
        inorder(root, &mut curr_smallest, &mut prev);
        curr_smallest
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::cell::RefCell;
    use std::rc::Rc;

    fn node(
        val: i32,
        left: Option<Rc<RefCell<TreeNode>>>,
        right: Option<Rc<RefCell<TreeNode>>>,
    ) -> Option<Rc<RefCell<TreeNode>>> {
        Some(Rc::new(RefCell::new(TreeNode { val, left, right })))
    }

    fn tree_to_string(root: Option<Rc<RefCell<TreeNode>>>) -> String {
        match root {
            Some(node) => {
                let node = node.borrow();
                format!(
                    "{}({},{})",
                    node.val,
                    tree_to_string(node.left.clone()),
                    tree_to_string(node.right.clone())
                )
            }
            None => "None".to_string(),
        }
    }

    #[test]
    fn main_asserts() {
        assert_eq!(
            Solution::min_diff_in_bst(node(
                4,
                node(2, node(1, None, None), node(3, None, None)),
                node(6, None, None)
            )),
            1
        );
        assert_eq!(
            Solution::min_diff_in_bst(node(
                1,
                node(0, None, None),
                node(48, node(12, None, None), node(49, None, None))
            )),
            1
        );
    }
}
